File: java/JavaStreams/src/test/java/pl/sebcel/livecoding/javastreams/employeeservice/GetMostFrequentEmployeesPerCompanyTests.java
package pl.sebcel.livecoding.javastreams.employeeservice;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.security.InvalidParameterException;
import java.util.List;
import java.util.Map;

import org.junit.jupiter.api.Test;

public class GetMostFrequentEmployeesPerCompanyTests {
	
	private EmployeeService cut = new EmployeeService();
	
	@Test
	public void should_throw_NullPointerException_when_companies_are_null() {
		assertThrows(NullPointerException.class, () -> cut.getMostFrequentEmployeesPerCompany(null));
	}
	
	@Test
	public void should_throw_NullPointerException_when_list_of_departments_for_a_company_is_null() {
		Company companyWithNullDepartments = new Company("A", null);
		assertThrows(NullPointerException.class, () -> {
			cut.getMostFrequentEmployeesPerCompany(List.of(companyWithNullDepartments));
		});
	}

	@Test
	public void should_throw_NullPointerException_when_list_of_employees_for_a_department_is_null() {
		Department departmentWithNullEmployees = new Department("Null", null);
		Company company = new Company("A", List.of(departmentWithNullEmployees));
		assertThrows(NullPointerException.class, () -> {
			cut.getMostFrequentEmployeesPerCompany(List.of(company));
		});
	}

	@Test
	public void should_return_data_for_all_companies() {
		Map<String, List<Employee>> result = cut.getMostFrequentEmployeesPerCompany(createCompanies());
		assertTrue(result.containsKey("A"));
		assertTrue(result.containsKey("B"));
	}

	@Test
	public void should_return_most_frequent_employees_per_company() {
		Map<String, List<Employee>> result = cut.getMostFrequentEmployeesPerCompany(createCompanies());
		assertEquals(1, result.get("A").size());
		assertTrue(result.get("A").stream().anyMatch(e -> e.name().equals("Alice")));
		
		assertEquals(2, result.get("B").size());
		List<String> employeeNames = result.get("B").stream().map(e -> e.name()).toList();
		assertTrue(employeeNames.contains("Tom"));
		assertTrue(employeeNames.contains("Jerry"));
	}

	
	private List<Company> createCompanies() {
		return List.of(
				com("A",
						dep("HR",
								emp("Alice"),
								emp("Bob"),
								emp("Charlie")),
						dep("IT", 
								emp("Alice"),
								emp("Bob")),
						dep("Sales", 
								emp("Alice"))),
				com("B",
						dep("Sales",
								emp("Tom"),
								emp("Jerry")),
						dep("HR",
								emp("Tom"),
								emp("Jerry")))
				);
	}
	
	private Employee emp(String name) {
		return new Employee(name);
	}
	
	private Department dep(String name, Employee... employees) {
		return new Department(name, List.of(employees));
	}

	private Company com(String name, Department... departments) {
		return new Company(name, List.of(departments));
	}

}

File: csharp/LINQtoObjects.Test/EmployeeService/GetMostFrequentEmployeesPerCompanyTests.cs
using LiveCodingExercises.LINQtoObjects.EmployeeService;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Xunit;

namespace LiveCodingExercises.LINQtoObjects.Test.EmployeeService
{
    public class GetMostFrequentEmployeesPerCompanyTests
    {
        private LINQtoObjects.EmployeeService.EmployeeService cut = new();

        [Fact]
        public void Should_return_data_for_all_companies()
        {
            var result = cut.GetMostFrequentEmployeesPerCompany(CreateCompanies());
            Assert.True(result.ContainsKey("A"));
            Assert.True(result.ContainsKey("B"));
        }

        [Fact]
        public void Should_return_most_frequent_employees_per_each_company()
        {
            var result = cut.GetMostFrequentEmployeesPerCompany(CreateCompanies());
            var mostFrequentEmployeesForCompanyA = result["A"].Select(e => e.Name).ToList();
            var mostFrequentEmployeesForCompanyB = result["B"].Select(e => e.Name).ToList();

            Assert.Single(mostFrequentEmployeesForCompanyA);
            Assert.Contains("Alice", mostFrequentEmployeesForCompanyA);

            Assert.Equal(2, mostFrequentEmployeesForCompanyB.Count);
            Assert.Contains("Tom", mostFrequentEmployeesForCompanyB);
            Assert.Contains("Jerry", mostFrequentEmployeesForCompanyB);
        }

        private IEnumerable<Company> CreateCompanies()
        {
            return [
                new Company("A", [
                    new Department("IT", [
                        new Employee("Alice"),
                        new Employee("Bob"),
                        new Employee("Charlie"),
                        ]),
                    new Department("HR", [
                        new Employee("Alice"),
                        new Employee("Bob"),
                        ]),
                    new Department("Sales", [
                        new Employee("Alice")
                        ])
                    ]),
                new Company("B", [
                    new Department("HR", [
                        new Employee("Tom"),
                        new Employee("Jerry")
                        ]),
                    new Department("Sales", [
                        new Employee("Tom"),
                        new Employee("Jerry")
                        ])
                ])
            ];
        }

    }
}
